File: 105.从前序与中序遍历序列构造二叉树.swift
/*
 * @lc app=leetcode.cn id=105 lang=swift
 *
 * [105] 从前序与中序遍历序列构造二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */
class Solution {
    func buildTree(_ preorder: [Int], _ inorder: [Int]) -> TreeNode? {
        var preorderIndex = 0
        guard preorder.count > 0, inorder.count > 0, preorder.count == inorder.count else {
            return nil
        }
        
        var map = [Int: Int]()
        for i in 0..<inorder.count{
            map[inorder[i]] = i
        }
        
        
        
        func buildTree(left: Int, right: Int) -> TreeNode?{
            guard left < right else {
                return nil
            }
            
            //获取根节点
            let node = TreeNode.init(preorder[preorderIndex])
            //生成左右子树
            let rootIndex = map[preorder[preorderIndex]]!
            preorderIndex += 1
            node.left = buildTree(left: left, right: rootIndex)
            node.right = buildTree(left: rootIndex + 1, right: right)
            return node
        }
        
        return buildTree(left: 0, right: preorder.count)
        
    }
}
// @lc code=end

File: 105.从前序与中序遍历序列构造二叉树.py
#
# @lc app=leetcode.cn id=105 lang=python3
#
# [105] 从前序与中序遍历序列构造二叉树
#

# @lc code=start
# Definition for a binary tree node.
# from typing import List

# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        if len(preorder) == 0:
            return None
        
        if len(preorder) == 1:
            assert len(inorder) == 1
            return TreeNode(preorder[0])
        
        rootValue = preorder[0]
        root = TreeNode(rootValue)
        
        rootIndexInorder = inorder.index(rootValue)
        
        leftInOrder = inorder[0:rootIndexInorder]
        leftPreorder = preorder[1:rootIndexInorder + 1]
        root.left = self.buildTree(leftPreorder, leftInOrder)
        
        rightInOrder = inorder[rootIndexInorder + 1:]
        rightPreorder = preorder[rootIndexInorder + 1:]
        root.right = self.buildTree(rightPreorder, rightInOrder)

        return root
    
# result = Solution().buildTree([3,9,20,15,7], [9,3,15,20,7])        
# print(result)        
# @lc code=end

